src/base/searchengine.h

```diff
@@ -38,10 +40,119 @@
 class QProcess;
 class QTimer;
 
+template <class T, std::size_t N>
+class Version
+{
+public:
+    typedef T ComponentType;
+    typedef Version<T,N> this_type;
+
+    Version()
+    {
+        m_components.fill(T());
+    }
+
+    template <class ...Other>
+    Version(Other... components)
+        : m_components {{components...}}
+    {
+    }
+
+    Version(const QString &version)
+        : Version {version.split(QLatin1Char('.'))}
+    {
+    }
+
+    Version(const QByteArray &version)
+        : Version {version.split('.')}
+    {
+    }
+
+    ComponentType majorNumber() const {
+        static_assert(N >= 1, "Array size is too small");
+        return m_components[0];
+    }
+
+    ComponentType minorNumber() const {
+        static_assert(N >= 2, "Array size is too small");
+        return m_components[1];
+    }
+
+    ComponentType revisionNumber() const {
+        static_assert(N >= 2, "Array size is too small");
+        return m_components[2];
+    }
+
+    ComponentType patchNumber() const {
+        static_assert(N >= 3, "Array size is too small");
+        return m_components[3];
+    }
+
+    ComponentType operator[](std::size_t i) const
+    {
+        return m_components.at(i);
+    }
+
+    QString toString() const
+    {
+        QString res = QString::number((*this)[0]);
+        for (std::size_t i = 1; i < N; ++i) {
+            res += QLatin1Char('.') + QString::number((*this)[i]);
+        }
+        return res;
+    }
+
+    bool operator == (const this_type &other) const
+    {
+        return m_components == other.m_components;
+    }
+
+    bool operator != (const this_type &other) const
+    {
+        return !(*this == other);
+    }
+
+    bool operator < (const this_type &other) const
+    {
+        for (std::size_t i = 0; i < N - 1; ++i) {
+            if (m_components[i] > other.m_components[i])
+                return false;
+        }
+        return m_components.back() < other.m_components.back();
+    }
+
+    bool operator > (const this_type &other) const
+    {
+        for (std::size_t i = 0; i < N - 1; ++i) {
+            if (m_components[i] < other.m_components[i])
+                return false;
+        }
+        return m_components.back() > other.m_components.back();
+    }
+private:
+    template <class QtStringsList>
+    Version (const QtStringsList &versionParts)
+    {
+        if (versionParts.size() != N)
+            throw std::logic_error("Incorrect number of version components");
+        bool ok = false;
+        for (std::size_t i = 0; i < N; ++i) {
+            m_components[i] = static_cast<T>(versionParts[i].toInt(&ok));
+            if (!ok)
+                throw std::runtime_error("Can not parse version components");
+        }
+    }
+    using ComponentsArray = std::array<T, N>;
```

Comment:
Separate declarations of different type (ie, methods, types, fields) by blank line.